### PR TITLE
feat: MFE core web-vitals detection

### PR DESCRIPTION
**Pull request by metal-messiah**
Add internal detection methods for FCP, INP, CLS, and LCP when spawned by a MFE
---

### Overview

This PR implements comprehensive Web Vitals tracking for Micro Frontend (MFE) components, adding detection and metadata collection for four Core Web Vitals metrics: First Contentful Paint (FCP), Largest Contentful Paint (LCP), Cumulative Layout Shift (CLS), and Interaction to Next Paint (INP).

#### Key Changes

**Unified Vitals Object Structure**
Each vital metric is tracked as a unified object containing both the measurement value and rich contextual metadata. These objects are reported directly as JSON in the `MicroFrontEndTiming` event:

```javascript
{
  'nr.vitals.fcp': {
    value: 1234.56,      // Timestamp when FCP occurred
    loadState: 'loading'  // Document readyState at FCP
  },
  'nr.vitals.lcp': {
    value: 2345.67,      // Timestamp when LCP occurred
    size: 480000,        // Element size in pixels (width × height)
    elTag: 'IMG',        // HTML tag name of LCP element
    eid: 'hero-image',   // Element ID (if present)
    elUrl: 'https://...' // Element URL (images/videos/background-images)
  },
  'nr.vitals.cls': {
    value: 0.123,                    // Cumulative layout shift score
    largestShiftValue: 0.05,         // Value of largest individual shift
    largestShiftTime: 1234,          // Timestamp of largest shift
    largestShiftTarget: 'div#hero',  // CSS selector of element that shifted
    loadState: 'complete'            // Document readyState when measured
  },
  'nr.vitals.inp': {
    value: 250,                      // Total interaction latency (ms)
    interactionTarget: 'button#submit', // CSS selector of interacted element
    interactionTime: 1234,           // Timestamp of interaction
    interactionType: 'click',        // Interaction type (click, keydown, etc.)
    inputDelay: 10,                  // Input delay (ms)
    processingDuration: 190,         // Event processing time (ms)
    presentationDelay: 50,           // Presentation delay (ms)
    nextPaintTime: 1484,             // Next paint timestamp
    loadState: 'interactive'         // Document readyState during interaction
  }
}
```

**MicroFrontEndTiming Event Attributes**
When an MFE deregisters (via `deregister()` API), a `MicroFrontEndTiming` event is generated with the following JSON object attributes:

**Important: these attributes are prefixed with nr. to ensure they are hidden while being evaluated**

**`nr.vitals.fcp` (object or null):**
- `value` - Time from script start to FCP (milliseconds)
- `loadState` - Document state when FCP occurred ('loading', 'interactive', or 'complete')

**`nr.vitals.lcp` (object or null):**
- `value` - Time from script start to LCP (milliseconds)
- `elTag` - HTML tag of the LCP element
- `eid` - Element ID attribute (if present)
- `elUrl` - URL for image/video elements or background images
- `size` - Element dimensions in pixels (width × height)

**`nr.vitals.cls` (object or null):**
- `value` - Cumulative layout shift score (0-1+)
- `largestShiftTarget` - CSS selector of element causing largest shift
- `largestShiftTime` - Timestamp of largest individual shift
- `largestShiftValue` - Value of largest individual shift
- `loadState` - Document state when CLS was measured

**`nr.vitals.inp` (object or null):**
- `value` - Worst interaction latency in milliseconds
- `interactionTarget` - CSS selector of interacted element
- `interactionTime` - Timestamp when interaction occurred
- `interactionType` - Type of interaction (click, keydown, keypress, etc.)
- `inputDelay` - Delay before event processing began (ms)
- `processingDuration` - Duration of event handler execution (ms)
- `presentationDelay` - Delay from processing to paint (ms)
- `nextPaintTime` - Timestamp when paint occurred
- `loadState` - Document state during interaction

```mermaid
sequenceDiagram
    participant App as MFE Application
    participant API as Register API
    participant Tracker as trackMFEVitals
    participant MO as MutationObserver
    participant PO as PerformanceObserver
    participant Events as Event Harvester

    App->>API: newrelic.register({ id, name })
    API->>Tracker: trackMFEVitals(mfeId)
    
    Note over Tracker: Initialize vitals object<br/>{ fcp, lcp, cls, inp } = null
    
    Tracker->>MO: Create FCP Observer
    Note over MO: Watches for first<br/>contentful paint
    Tracker->>MO: Create LCP Observer
    Note over MO: Watches for largest<br/>contentful paint
    Tracker->>PO: Create CLS Observer
    Note over PO: Watches for layout<br/>shift entries
    Tracker->>PO: Create INP Observer
    Note over PO: Watches for interaction<br/>event entries

    par Vitals Detection
        MO-->>Tracker: FCP detected
        Note over Tracker: vitals.fcp = {<br/>value, loadState<br/>}
        MO-->>Tracker: LCP detected (updates on larger elements)
        Note over Tracker: vitals.lcp = {<br/>value, size, elTag,<br/>eid, elUrl<br/>}
        PO-->>Tracker: CLS shift detected
        Note over Tracker: vitals.cls = {<br/>value, largestShiftValue,<br/>largestShiftTime,<br/>largestShiftTarget, loadState<br/>}
        PO-->>Tracker: INP interaction detected
        Note over Tracker: vitals.inp = {<br/>value, interactionTarget,<br/>interactionTime, interactionType,<br/>inputDelay, processingDuration,<br/>presentationDelay,<br/>nextPaintTime, loadState<br/>}
    end

    App->>API: mfe.deregister()
    API->>Tracker: Get vitals
    
    Tracker->>API: Return vitals object
    API->>Events: Send MicroFrontEndTiming event
    Note over Events: Vitals sent as JSON strings:<br/>"nr.vitals.fcp": "{...}"<br/>"nr.vitals.lcp": "{...}"<br/>"nr.vitals.cls": "{...}"<br/>"nr.vitals.inp": "{...}"
```

### Testing

**Unit Tests** (`tests/unit/common/v2/mfe-vitals.test.js`)
- ✅ FCP tracking with loadState metadata
- ✅ LCP tracking with element metadata (tag, id, url, size)
- ✅ LCP updates when larger elements appear
- ✅ CLS tracking with largest shift metadata
- ✅ INP tracking with full interaction timeline
- ✅ Element scope validation (only tracks within MFE boundary)
- ✅ Observer disconnect functionality
- ✅ Edge cases (missing IDs, SVG className objects, invalid elements)

**E2E Tests** (`tests/specs/api/register/mfe-vitals.e2e.js`)
- ✅ Real browser FCP/LCP capture with JSON object validation
- ✅ Metadata presence validation for all vitals
- ✅ Value sanity checks (reasonable ranges, relative ordering)
- ✅ Browser compatibility filters (Chromium-only tests for CLS/INP/LCP)
- ✅ Multiple MFE isolation testing
- ✅ Observer lifecycle and disconnection validation

**Browser Compatibility:**
- **FCP**: All browsers supporting First Contentful Paint
- **LCP**: Chromium-based browsers (Chrome, Edge, Opera)
- **CLS**: Chromium-based browsers (layout-shift PerformanceObserver)
- **INP**: Chromium-based browsers with PerformanceEventTiming API

**Manual Testing Checklist:**
- [ ] Verify MicroFrontEndTiming events are harvested with `nr.vitals.*` JSON objects
- [ ] Confirm objects are null when vitals are unsupported/not captured
- [ ] Test multiple MFEs on same page don't cross-contaminate vitals
- [ ] Verify JSON object structure is preserved in event payloads
- [ ] Check that deregister() properly disconnects observers and captures final values

### CODE REVIEW

File: src/common/v2/mfe-vitals.js [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+      }
+    } catch (e) {
+      // Element may be detached from DOM
+    }
+  })
+  observers.push(lcpObs)
+
+  // Track CLS - cumulative layout shift
+  let clsValue = 0
+  const clsObs = observePerformance(observers, { type: 'layout-shift', buffered: true }, (entry) => {
```
Comment (ptang-nr) [inline src/common/v2/mfe-vitals.js:179]: Asking to increase my understanding - why is `buffered: true` used for CLS but not LCP?

File: src/common/v2/mfe-vitals.js [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+      lcpObs?.disconnect()
+    } catch (e) {
+      // Observer may already be disconnected
+    }
+  }
+  ;['click', 'keydown', 'scroll'].forEach(type => {
+    globalScope.addEventListener(type, disconnectLCP, { once: true, passive: true })
+  })
+
+  // Disconnect all observers on visibility change or page unload
```
Comment (ptang-nr) [inline src/common/v2/mfe-vitals.js:254]: Why do we stop tracking CLS + INP once user focus leaves the page? (web-vitals seems to keep tracking when user returns)

File: src/common/v2/mfe-vitals.js (hunk 1 above)
Comment (metal-messiah) [reply src/common/v2/mfe-vitals.js:179]: LCP in this case doesn't use `buffered: true` because it's using a MutationObserver, not a PerformanceObserver. The `buffered` parameter only applies to PerformanceObservers.

We had to create a custom implementation to track content size in the portion of the DOM tree manually rather than using the browser's native largest-contentful-paint PerformanceEntry.

File: src/common/v2/mfe-vitals.js (hunk 2 above)
Comment (metal-messiah) [reply src/common/v2/mfe-vitals.js:254]: This was mostly an effort to minimize the processing footprint that MFE will need to take up.  We could revisit this though and change it to `pagehide` only if that feels more reliable

Comment (ptang-nr) [reply src/common/v2/mfe-vitals.js:254]: I was mostly just curious about the difference.  Sounds like it's a design choice for minimizing ongoing observations for MFEs, good to know!

File: src/common/v2/mfe-vitals.js [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    }
+  })
+  observers.push(lcpObs)
+
+  // Track CLS - cumulative layout shift
+  // Initialize CLS to 0 if browser supports it AND the MFE container exists in the DOM
+  let clsValue = 0
+  const clsObs = observePerformance(observers, { type: 'layout-shift', buffered: true }, (entry) => {
+    if (entry.hadRecentInput || !vitals.cls) return
+    (entry.sources || []).some(source => {
```
Comment (ptang-nr) [inline src/common/v2/mfe-vitals.js:181]: Minor suggestion: Maybe consider using `find()` to achieve similar result/return only the first matching result in a given MFE.  (This started as a "style" suggestion, but thought may be worth mentioning, as the code may be slightly shorter.)

**Comment by ptang-nr** [comment]
I also saw LCP and INP are newly supported in Firefox and Safari.  Might be good to mention in the PR notes.

**Review by ptang-nr — APPROVED** [review]

**Comment by metal-messiah** [comment]
will have another followup PR to refine the output after evaluation and API review, to expose valuable attributes and remove unneeded ones
